sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -477,5 +486,82 @@ object StateStore extends Logging {
+      providerId: StateStoreProviderId,
+      hadoopConf: Configuration) {
+
+    private val storeCpLocation = providerId.storeId.storeCheckpointLocation()
+    private val fm = CheckpointFileManager.create(storeCpLocation, hadoopConf)
+    private val schemaFileLocation = schemaFile(storeCpLocation)
+
+    fm.mkdirs(schemaFileLocation.getParent)
+
+    def check(keySchema: StructType, valueSchema: StructType): Unit = {
+      if (fm.exists(schemaFileLocation)) {
+        logDebug(s"Schema file for provider $providerId exists. Comparing with provided schema.")
+        val (storedKeySchema, storedValueSchema) = readSchemaFile()
+
+        def typesEq(schema1: StructType, schema2: StructType): Boolean = {
+          val fieldToType: StructField => (DataType, Boolean) = f => (f.dataType, f.nullable)
+          (schema1.length == schema2.length) &&
+            schema1.map(fieldToType).equals(schema2.map(fieldToType))
+        }
+
+        def namesEq(schema1: StructType, schema2: StructType): Boolean = {
+          schema1.fieldNames.toList.equals(schema2.fieldNames.toList)
+        }
+
+        val errorMsg = "Provided schema doesn't match to the schema for existing state! " +
+          "Please note that Spark allow difference of field name: check count of fields " +
+          "and data type of each field.\n" +
+          s"- provided schema: key $keySchema value $valueSchema\n" +
+          s"- existing schema: key $storedKeySchema value $storedValueSchema\n"
+
+        if (!typesEq(keySchema, storedKeySchema) || !typesEq(valueSchema, storedValueSchema)) {
+          logError(errorMsg)
+          throw new IllegalStateException(errorMsg)
+        } else if (!namesEq(keySchema, storedKeySchema) ||
+          !namesEq(valueSchema, storedValueSchema)) {
+          logInfo("Detected the change of field name, will overwrite schema file to new.")
+          // It tries best-effort to overwrite current schema file.
+          // the schema validation doesn't break even it fails, though it might miss on detecting
+          // change of field name which is not a big deal.
+          createSchemaFile(keySchema, valueSchema)
+        }
+      } else {
+        // schema doesn't exist, create one now
+        logDebug(s"Schema file for provider $providerId doesn't exist. Creating one.")
+        createSchemaFile(keySchema, valueSchema)
+      }
+    }
+
+    private def readSchemaFile(): (StructType, StructType) = {
+      val inStream = fm.open(schemaFileLocation)
+      try {
+        val keySchemaStr = inStream.readUTF().stripSuffix("\n")
+        val valueSchemaStr = inStream.readUTF().stripSuffix("\n")
+
+        (StructType.fromString(keySchemaStr), StructType.fromString(valueSchemaStr))
+      } finally {
+        inStream.close()
+      }
+    }
+
+    private def createSchemaFile(keySchema: StructType, valueSchema: StructType): Unit = {
+      val outStream = fm.createAtomic(schemaFileLocation, overwriteIfPossible = true)
```

Comment:
Nit: `overwriteIfPossible` keyword not needed.

Reply:
Spark project tends to prefer adding name for boolean parameter. Please search for `overwriteIfPossible = `.